src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -3056,7 +3091,179 @@ public static void Main()
+  IL_0005:  call       ""System.ValueTuple<int, int>..ctor(int, int)""
+  IL_000a:  ldloca.s   V_1
+  IL_000c:  ldc.i4.1
+  IL_000d:  ldc.i4.2
+  IL_000e:  call       ""System.ValueTuple<int, int>..ctor(int, int)""
+  IL_0013:  br.s       IL_0015
+  IL_0015:  br.s       IL_0017
+  IL_0017:  ldloc.0
+  IL_0018:  ldfld      ""int System.ValueTuple<int, int>.Item1""
+  IL_001d:  ldloc.1
+  IL_001e:  ldfld      ""int System.ValueTuple<int, int>.Item1""
+  IL_0023:  bne.un.s   IL_0035
+  IL_0025:  ldloc.0
+  IL_0026:  ldfld      ""int System.ValueTuple<int, int>.Item2""
+  IL_002b:  ldloc.1
+  IL_002c:  ldfld      ""int System.ValueTuple<int, int>.Item2""
+  IL_0031:  ceq
+  IL_0033:  br.s       IL_0036
+  IL_0035:  ldc.i4.0
+  IL_0036:  call       ""void System.Console.Write(bool)""
+  IL_003b:  nop
+  IL_003c:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TestEqualOnNullableVsNullableTuples_OneSideNeverNull()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        System.Console.Write(((int, int)?) (1, 2) == (1, 2));
+    }
+}
+";
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "True");
+            verifier.VerifyIL("C.Main", @"{
+  // Code size       42 (0x2a)
+  .maxstack  3
+  .locals init (System.ValueTuple<int, int> V_0)
+  IL_0000:  nop
+  IL_0001:  ldloca.s   V_0
+  IL_0003:  ldc.i4.1
+  IL_0004:  ldc.i4.2
+  IL_0005:  call       ""System.ValueTuple<int, int>..ctor(int, int)""
+  IL_000a:  br.s       IL_000c
+  IL_000c:  br.s       IL_000e
+  IL_000e:  ldloc.0
+  IL_000f:  ldfld      ""int System.ValueTuple<int, int>.Item1""
+  IL_0014:  ldc.i4.1
+  IL_0015:  bne.un.s   IL_0022
+  IL_0017:  ldloc.0
+  IL_0018:  ldfld      ""int System.ValueTuple<int, int>.Item2""
+  IL_001d:  ldc.i4.2
+  IL_001e:  ceq
+  IL_0020:  br.s       IL_0023
+  IL_0022:  ldc.i4.0
+  IL_0023:  call       ""void System.Console.Write(bool)""
+  IL_0028:  nop
+  IL_0029:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TestEqualOnNullableVsNullableTuples_TupleAlwaysNull()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        System.Console.Write(((int, int)?)null == ((int, int)?)null);
+    }
+}
+";
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "True");
+            verifier.VerifyIL("C.Main", @"{
+  // Code size      110 (0x6e)
+  .maxstack  3
+  .locals init (bool V_0,
+                System.ValueTuple<int, int> V_1,
+                System.ValueTuple<int, int> V_2,
+                (int, int)? V_3)
+  IL_0000:  nop
+  IL_0001:  ldloca.s   V_3
+  IL_0003:  dup
+  IL_0004:  initobj    ""(int, int)?""
+  IL_000a:  call       ""bool (int, int)?.HasValue.get""
```

Comment:
>""bool (int, int)?.HasValue.get"" [](start = 23, length = 33)

I'd expect to see much simpler IL.

Reply:
Like for the test below

---
In reply to: [176297764](https://github.com/dotnet/roslyn/pull/25644#discussion_r176297764) [](ancestors = 176297764)